src/trace/trace.c

```diff
@@ -96,19 +116,144 @@ static inline void mtrace_event(const char *data, uint32_t length)
 	platform_shared_commit(trace, sizeof(*trace));
 }
 
+#if CONFIG_TRACE_FILTERING_VERBOSITY
 /**
- * \brief Runtime trace filtering
+ * \brief Runtime trace filtering based on verbosity level
  * \param lvl log level (LOG_LEVEL_ ERROR, INFO, DEBUG ...)
  * \param uuid uuid address
  * \return false when trace is filtered out, otherwise true
  */
-static inline bool trace_filter_pass(uint32_t lvl,
-				     const struct tr_ctx *ctx)
+static inline bool trace_filter_verbosity(uint32_t lvl, const struct tr_ctx *ctx)
 {
 	STATIC_ASSERT(LOG_LEVEL_CRITICAL < LOG_LEVEL_VERBOSE,
 		      LOG_LEVEL_CRITICAL_MUST_HAVE_LOWEST_VALUE);
 	return lvl <= ctx->level;
 }
+#endif /* CONFIG_TRACE_FILTERING_VERBOSITY */
+
+#if CONFIG_TRACE_FILTERING_ADAPTIVE
+static void prepare_to_emit(struct recent_log_entry *entry)
+{
+	struct trace *trace = trace_get();
+	struct recent_log_entry *entry_to_emit =
+		&trace->trace_core_context[cpu_get_id()].entry_to_emit;
+
+	memcpy_s(entry_to_emit, sizeof(*entry_to_emit), entry, sizeof(*entry));
+
+	platform_shared_commit(trace, sizeof(*trace));
+
+	memset(entry, 0, sizeof(*entry));
+}
+
+static void emit_recent_entry(void)
+{
+	struct trace *trace = trace_get();
+	struct recent_log_entry *entry_to_emit =
+		&trace->trace_core_context[cpu_get_id()].entry_to_emit;
+
+	_log_message(trace_log_unfiltered, false, LOG_LEVEL_INFO, _TRACE_INV_CLASS, &dt_tr,
+		     _TRACE_INV_ID, _TRACE_INV_ID, "Suppressed %u similar messages: %pE",
+		     entry_to_emit->suppress_count, entry_to_emit->entry_id);
+
+	memset(entry_to_emit, 0, sizeof(*entry_to_emit));
+
+	platform_shared_commit(trace, sizeof(*trace));
+}
+
+static void emit_recent_entries(uint64_t current_ts)
+{
+	struct trace *trace = trace_get();
+	struct recent_log_entry *entry_to_emit =
+		&trace->trace_core_context[cpu_get_id()].entry_to_emit;
+	struct recent_log_entry *recent_entries =
+		trace->trace_core_context[cpu_get_id()].recent_entries;
+	int i;
+
+	if (entry_to_emit->entry_id)
+		emit_recent_entry();
+
+	for (i = 0; i < CONFIG_TRACE_RECENT_ENTRIES_COUNT; i++) {
```

Comment:
also an inline comment here explaining what we are doing here.

Reply:
done